File: .temp/flutter_app/lib/data/local/dao/category_dao.dart
import 'package:drift/drift.dart';
import '../database.dart';

part 'category_dao.g.dart';

@DriftAccessor(tables: [Categories])
class CategoryDao extends DatabaseAccessor<AppDatabase> with _$CategoryDaoMixin {
  CategoryDao(super.db);

  Stream<List<CategoryEntity>> getAllCategories() {
    return (select(categories)
          ..orderBy([
            (t) => OrderingTerm(expression: t.orderIndex, mode: OrderingMode.asc),
            (t) => OrderingTerm(expression: t.id, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Future<int> insertCategory(Insertable<CategoryEntity> category) {
    return into(categories).insert(category, mode: InsertMode.insertOrIgnore);
  }

  Future<bool> updateCategory(Insertable<CategoryEntity> category) {
    return update(categories).replace(category);
  }

  Future<void> updateCategories(List<Insertable<CategoryEntity>> categoryList) async {
    await batch((batch) {
      batch.replaceAll(categories, categoryList);
    });
  }

  Future<int> deleteCategory(Insertable<CategoryEntity> category) {
    return delete(categories).delete(category);
  }

  Future<CategoryEntity?> getCategoryById(int id) {
    return (select(categories)..where((t) => t.id.equals(id))).getSingleOrNull();
  }

  Future<int> getCount() async {
    final countExp = categories.id.count();
    final query = selectOnly(categories)..addColumns([countExp]);
    final result = await query.getSingle();
    return result.read(countExp) ?? 0;
  }
}

File: shared/src/commonMain/kotlin/com/sans/finance/data/local/dao/CategoryDao.kt
package com.sans.finance.data.local.dao

import androidx.room3.Dao
import androidx.room3.Delete
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import androidx.room3.Update
import com.sans.finance.data.local.entity.CategoryEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface CategoryDao {
    @Query("SELECT * FROM categories WHERE type = :type ORDER BY orderIndex ASC, id ASC")
    fun getCategoriesByType(type: String): Flow<List<CategoryEntity>>

    @Query("SELECT * FROM categories ORDER BY orderIndex ASC, id ASC")
    fun getAllCategories(): Flow<List<CategoryEntity>>

    @Query("SELECT * FROM categories")
    suspend fun getAllCategoriesSync(): List<CategoryEntity>

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertCategory(category: CategoryEntity)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateCategory(category: CategoryEntity)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateCategories(categories: List<CategoryEntity>)

    @Delete
    suspend fun deleteCategory(category: CategoryEntity)

    @Query("SELECT * FROM categories WHERE id = :id")
    suspend fun getCategoryById(id: Long): CategoryEntity?

    @Query("SELECT COUNT(*) FROM categories")
    suspend fun getCount(): Int
}
